src/System.Net.Http/tests/UnitTests/MockContent.cs

```diff
@@ -0,0 +1,173 @@
+        DontOverrideCreateContentReadStream = 0x8,
+        CanCalculateLength = 0x10,
+        ThrowInTryComputeLength = 0x20,
+        ThrowInAsyncSerializeMethods = 0x40
+    }
+
+    public class MockException : Exception
+    {
+        public MockException() { }
+        public MockException(string message) : base(message) { }
+        public MockException(string message, Exception inner) : base(message, inner) { }
+    }
+
+    public class MockContent : HttpContent
+    {
+        private byte[] _mockData;
+        private MockOptions _options;
+        private Exception _customException;
+
+        public int TryComputeLengthCount { get; private set; }
+        public int SerializeToStreamAsyncCount { get; private set; }
+        public int CreateContentReadStreamCount { get; private set; }
+        public int DisposeCount { get; private set; }
+
+        public byte[] MockData
+        {
+            get { return _mockData; }
+        }
+
+        public MockContent()
+            : this((byte[])null, MockOptions.None)
+        {
+        }
+
+        public MockContent(byte[] mockData)
+            : this(mockData, MockOptions.None)
+        {
+        }
+
+        public MockContent(MockOptions options)
+            : this((byte[])null, options)
+        {
+        }
+
+        public MockContent(Exception customException, MockOptions options)
+            : this((byte[])null, options)
+        {
+            _customException = customException;
+        }
+
+        public MockContent(byte[] mockData, MockOptions options)
+        {
+            _options = options;
+
+            if (mockData == null)
+            {
+                _mockData = Encoding.UTF8.GetBytes("data");
+            }
+            else
+            {
+                _mockData = mockData;
+            }
+        }
+
+        public byte[] GetMockData()
+        {
+            return _mockData;
+        }
+
+        protected internal override bool TryComputeLength(out long length)
+        {
+            TryComputeLengthCount++;
+
+            if ((_options & MockOptions.ThrowInTryComputeLength) != 0)
+            {
+                throw new MockException();
+            }
+
+            if ((_options & MockOptions.CanCalculateLength) != 0)
+            {
+                length = _mockData.Length;
+                return true;
+            }
+            else
+            {
+                length = 0;
+                return false;
+            }
+        }
+
+        protected override Task SerializeToStreamAsync(Stream stream, TransportContext context)
+        {
+            SerializeToStreamAsyncCount++;
+
+            if ((_options & MockOptions.ReturnNullInCopyToAsync) != 0)
+            {
+                return null;
+            }
+
+            if ((_options & MockOptions.ThrowInAsyncSerializeMethods) != 0)
+            {
+                throw _customException;
+            }
+
+            return Task.Factory.StartNew(() =>
```

Comment:
I'd suggest switching this to be Task.Run, which will always target TaskScheduler.Default (the thread pool) rather than Task.Factory.StartNew, which unless overridden will target the current scheduler.  The latter can cause problems if the environment has a non-default scheduler configured, that scheduler has a max set on th enumber of concurrent operations allowed, and some code somewhere blocks waiting for one of these operations to complete, as that can lead to deadlock.  We've had issues with this sporadically due to xunit's usage of TaskScheduler; xunit's implementation has been fixed, but we haven't yet picked up that fix in corefx.